### ISSUE TITLE
[sklearn#34270] Declare metadata properties to be used by the routers

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#340]
**Mirrored from [scikit-learn#34270](https://github.com/scikit-learn/scikit-learn/issues/34270)**

Original title: Declare metadata properties to be used by the routers
Original labels: Needs Decision, Metadata Routing
Created: 2026-06-12T14:24:25Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

When working on https://github.com/scikit-learn/scikit-learn/pull/34263, I realized that `Pipeline` may not be the only place where we need to specify metadata that needs to be transformed like `X`.

There are several estimators that transform `X` internally (e.g. KMeans centers `X`). Now that they're going to become routers, they'll need to be able to transform some metadata, like `X_val`. I would like to avoid introducing `transform_input` in all these estimators.

On a related topic, meta-estimators with a CV split the metadata like `X` when it has the same length. This is very brittle and a well known issue. For instance if you pass a `X_val` that has the same size a `X` to a grid search, it will be split, and thus the callback won't receive the full `X_val`.

What's currently missing from metadata-routing to gracefully handle these 2 uses cases is a way to declare some properties for the metadata that you request. Discussing with @FrancoisPgm we've thought about a few alternatives:

- Add the possibility to declare additional property at request time. Something like:
  ```py
  LogisticRegression().set_fit_request(sample_weight=(True, MetadataProperties(split_like_X=True)))
  # or
  HistGradientBoostingClassifier().set_fit_request(X_val=(True, MetadataProperties(transform_like_X=True)))
  ```

  It adds more verbosity and burden on the user.

- Declare it only once using a kind of global registry, e.g.
  ```py
  sklearn.metadata_properties(sample_weight=MetadataProperties(split_like_X=True))
  ```

  Still adds some verbosity, except that it can be done once and not in all requests. The burden is still on the user side.

- Declare the property at the estimator level, maybe as a tag, maybe something else:
  ```py
  class LogisticRegression:

      metadata_properties = {
          "sample_weight": MetadataProperties(split_like_X=True)
      }

      def fit(X, y, sample_weight):
          ...
  ```

  The burden is on the developer side. The property is defined by the consumer that knows what it expects. No extra verbosity on the user side.

where MetadataProperties is a small dataclass like
```py
  @dataclass
  class MetadataProperties
      split_like_X: bool = False
      transform_like_X: bool = False
```

Writing this I realized that the 3rd option really feels more appropriate. I'm leaving the first 2 options for comparison. Ping @adrinjalali @StefanieSenger @ogrisel @FrancoisPgm what do you think ?

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/83 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
